lib/services/discourse.ts

```diff
@@ -17,211 +17,80 @@
-import { Messenger } from './messenger';
+import { RequestPromiseOptions } from 'request-promise';
 import {
-	MessengerAction, MessengerEmitResponse, MessengerEvent, ReceiptContext, TransmitContext
-} from './messenger-types';
+	DiscourseConnectionDetails, DiscourseEmitContext, DiscourseEvent, DiscourseHandle,
+	DiscourseResponse
+} from './discourse-types';
 import { ServiceEmitter, ServiceListener } from './service-types';
+import { ServiceUtilities } from './service-utilities';
 
-export class DiscourseService extends Messenger implements ServiceListener, ServiceEmitter {
+export class DiscourseService extends ServiceUtilities<string> implements ServiceListener, ServiceEmitter {
 	private static _serviceName = path.basename(__filename.split('.')[0]);
 	// There are circumstances in which the discourse web-hook will fire twice for the same post, so track.
 	private postsSynced = new Set<number>();
-	private data: DiscourseConstructor;
-
-	public constructor(data: DiscourseConstructor, listen = true) {
-		super(listen);
-		this.data = data;
-	}
-
-	/**
-	 * Promise to turn the data enqueued into a generic message format.
-	 * @param data  Raw data from the enqueue, remembering this is as dumb and quick as possible.
-	 * @returns     A promise that resolves to the generic form of the event.
-	 */
-	public makeGeneric = (data: MessengerEvent): Promise<ReceiptContext> => {
-		// Encode once the common parts of a request
-		const getGeneric = {
-			json: true,
-			method: 'GET',
-			qs: {
-				api_key: this.data.token,
-				api_username: this.data.username,
-			},
-			// appended before execution
-			uri: `https://${this.data.instance}`,
-		};
-		// Gather more complete details of the enqueued event
-		const getPost = _.cloneDeep(getGeneric);
-		getPost.uri += `/posts/${data.rawEvent.id}`;
-		const getTopic = _.cloneDeep(getGeneric);
-		getTopic.uri += `/t/${data.rawEvent.topic_id}`;
-		return Promise.props({
-			post: request(getPost),
-			topic: request(getTopic),
-		})
-		.then((details: {post: any, topic: any}) => {
-			// Gather metadata and resolve
-			const metadata = Messenger.extractMetadata(details.post.raw);
-			const first = details.post.post_number === 1;
-			return {
-				action: MessengerAction.Create,
-				first,
-				genesis: metadata.genesis || data.source,
-				// post_type 4 seems to correspond to whisper
-				hidden: first ? !details.topic.visible : details.post.post_type === 4,
-				source: DiscourseService._serviceName,
-				sourceIds: {
-					// These come in as integers, but should be strings
-					flow: details.topic.category_id.toString(),
-					message: details.post.id.toString(),
-					thread: details.post.topic_id.toString(),
-					url: getTopic.uri,
-					user: details.post.username,
-				},
-				text: metadata.content,
-				title: details.topic.title,
-			};
-		});
+	private connectionDetails: DiscourseConnectionDetails;
+
+	constructor(data: DiscourseConnectionDetails, listen: boolean) {
```

Comment:
Or, it would seem, to sleep on it.  I woke up this morning know how to approach this.  The constructor objects now have an optional boolean `deaf`, which if truthy skips the listen code.